upcloud/resource_upcloud_gateway_test.go

```diff
@@ -26,21 +30,116 @@ func TestAccUpcloudGateway(t *testing.T) {
+					resource.TestCheckResourceAttr(conn1Name, "remote_route.0.type", "static"),
+					resource.TestCheckResourceAttr(conn1Name, "remote_route.0.static_network", "100.123.123.0/24"),
+
+					resource.TestCheckResourceAttr(tunnel1Name, "name", "test-tunnel"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "connection_id"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "operational_state"),
+					resource.TestCheckResourceAttr(tunnel1Name, "local_address_name", "my-public-ip"),
+					resource.TestCheckResourceAttr(tunnel1Name, "remote_address", "100.123.123.10"),
+					resource.TestCheckResourceAttr(tunnel1Name, "ipsec_auth_psk.0.psk", "presharedkey1"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.child_rekey_time"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.dpd_delay"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.dpd_timeout"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.ike_lifetime"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.rekey_time"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.phase1_algorithms.0"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.phase1_dh_group_numbers.0"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.phase1_integrity_algorithms.0"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.phase2_algorithms.0"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.phase2_dh_group_numbers.0"),
+					resource.TestCheckResourceAttrSet(tunnel1Name, "ipsec_properties.0.phase2_integrity_algorithms.0"),
+
+					resource.TestCheckResourceAttr(conn2Name, "name", "test-connection2"),
+					resource.TestCheckResourceAttrSet(conn2Name, "gateway"),
+					resource.TestCheckResourceAttr(conn2Name, "type", "ipsec"),
+					resource.TestCheckResourceAttr(conn2Name, "local_route.#", "1"),
+					resource.TestCheckResourceAttr(conn2Name, "local_route.0.name", "local-route2"),
+					resource.TestCheckResourceAttr(conn2Name, "local_route.0.type", "static"),
+					resource.TestCheckResourceAttr(conn2Name, "local_route.0.static_network", "22.123.123.0/24"),
+					resource.TestCheckResourceAttr(conn2Name, "remote_route.#", "1"),
+					resource.TestCheckResourceAttr(conn2Name, "remote_route.0.name", "remote-route2"),
+					resource.TestCheckResourceAttr(conn2Name, "remote_route.0.type", "static"),
+					resource.TestCheckResourceAttr(conn2Name, "remote_route.0.static_network", "222.123.123.0/24"),
+
+					resource.TestCheckResourceAttr(tunnel2Name, "name", "test-tunnel2"), // Single check for tunnel2 just to make sure it gets created
+
+					// This field is deprecated, can be removed later
+					resource.TestCheckTypeSetElemNestedAttrs(name, "addresses.*", map[string]string{"name": "my-public-ip"}),
+				),
+			},
```

Comment:
Adding a new step with something like below would test the import as well:
```
			ResourceName:      name,
			ImportState:       true,
			ImportStateVerify: true,
```

Reply:
Good catch, done